torch/testing/_internal/common_methods_invocations.py

```diff
@@ -750,9 +758,165 @@ def default_test_dtypes(self, device_type):
+    optional for a reduction operator to implement, such as an identity value.
+
+    If a particular reduction operator does not yet implement parts of the API
+    that are required, this should be signaled by asserting that the relevant
+    tests fail so that it is clear that support should be added.
+
+    For ReductionOpInfo tests, a reduction operator is an operator that:
+        - accepts dim and keepdim parameters
+        - reduces one or more dimensions of a tensor to one or more values
+
+    NOTE
+    The API for reduction operators has not yet been finalized and some
+    requirements may change.
+
+    See tests in test/test_reductions.py
+    """
+
+    def __init__(
+        self, name, *,
+
+        # The identity value of the operator.
+        identity: Optional[Union[int, float, complex]] = None,
+
+        # possible values are:
+        # - propagate: NaN values are propagated to the output
+        # - omit: only non-NaN values participate in computing reduction
+        nan_policy: str = 'propagate',
+
+        # Whether the operator supports reducing multiple dimensions.
+        supports_multiple_dims: bool = True,
+
+        # Whether the operator promotes integral dtypes to floating point dtypes
+        promotes_int_to_float: bool = False,
+
+        # Whether the operator promotes all integral dtypes (including bool) to int64
+        promotes_int_to_int64: bool = False,
+
+        # If a specific dtype is given, then the operator always returns that
+        # dtype irrespective of the input dtype. If None, the operator returns
+        # the dtype according to the type promotion rules above.
+        result_dtype: Optional[torch.dtype] = None,
+
+        # A function that takes the input tensor and optional dim kwarg and
+        # generates tuples of args and kwargs to use when calling the operator
+        # and reference implementation with the given dim kwarg. Note that some
+        # tests will only use the first generated args and kwargs.
+        generate_args_kwargs: Callable = lambda t, dim=None: (yield tuple(), {}),
+
+        # Options from the OpInfo base class
+        **kwargs,
+    ):
+        # These are mutually exclusive options
+        assert not (promotes_int_to_float and promotes_int_to_int64)
+        assert not (result_dtype and promotes_int_to_float)
+        assert not (result_dtype and promotes_int_to_int64)
+
+        assert nan_policy in ('propagate', 'omit')
+
+        # ReductionOpInfo tests generate their own input tensors and the dim
+        # and keepdim parameters. Operators can provide a `generate_args_kwargs`
+        # to augment the arguments to the operator and the reference implementation
+        # or specify `sample_inputs_func` to provide their own sample inputs for
+        # the OpInfo base class. However, the test_reductions.py file will still
+        # generate its own inputs.
+        def sample_inputs_func(*args, **kwargs):
```

Comment:
I understand the comment but I'm not sure I understand why the structure of this function is necessary -- should the default sample inputs func just be a partial of one of the above funcs? Why the kwarg surgery here?

Reply:
This is necessary to augment the sample inputs from `sample_inputs_reduction` with the args and kwargs from `generate_args_kwargs`.

Reply:
I guess I can buy that. An alternative design might have had `sample_inputs_reduction` accept the `generate_args_kwargs` function as an input and perform this logic itself

Reply:
Yes, that need actually came up later when porting other ops so I might do that. Thanks for the suggestion.